algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
             .collect::<BTreeMap<_, _>>();
         end_timer!(v_X_at_beta_time);
 
-        let z_b_s_at_beta = z_b_s
-            .iter()
-            .map(|(circuit_id, z_b_i)| {
-                let z_b_i_s = z_b_i.iter().map(|z_b| evals.get_lc_eval(z_b, beta).unwrap()).collect::<Vec<F>>();
-                (*circuit_id, z_b_i_s)
-            })
-            .collect::<BTreeMap<_, _>>();
-        let batch_z_b_s_at_beta = z_b_s_at_beta
-            .iter()
-            .zip_eq(batch_combiners.values())
-            .zip_eq(r_alpha_at_beta_s.values())
-            .map(|(((circuit_id, z_b_i_at_beta), combiners), &r_alpha_at_beta)| {
-                let z_b_at_beta = z_b_i_at_beta
-                    .iter()
-                    .zip_eq(&combiners.instance_combiners)
-                    .map(|(z_b_at_beta, instance_combiner)| *z_b_at_beta * instance_combiner)
-                    .sum::<F>();
-                (*circuit_id, r_alpha_at_beta * z_b_at_beta)
-            })
-            .collect::<BTreeMap<_, _>>();
-        let g_1_at_beta = evals.get_lc_eval(&g_1, beta)?;
-
-        let combined_x_at_betas = state
+        let x_at_betas = state
             .circuit_specific_states
             .iter()
             .enumerate()
             .map(|(i, (circuit_id, circuit_state))| {
                 let lag_at_beta = circuit_state.input_domain.evaluate_all_lagrange_coefficients(beta);
```

Comment:
I think we can construct this once per `circuit_id`, instead of once per circuit_state

Reply:
Hmm not sure what you mean, `circuit_specific_states` contains `(circuit_id, circuit_state)`, it is already creating only one per `circuit_id`

Reply:
You're right, I misread the code.